File: cdn-proxy.js
const http = require('http');
const https = require('https');
const url = require('url');

const PORT = 8001;

const CDN_MAPPINGS = {
    '/wllama/': 'https://cdn.jsdelivr.net/npm/@wllama/wllama@2.3.1/dist/',
    '/coi-serviceworker/': 'https://cdn.jsdelivr.net/npm/coi-serviceworker@0.1.9/'
};

function forwardRequest(req, res, targetUrl) {
    const options = url.parse(targetUrl);
    options.headers = {
        'User-Agent': req.headers['user-agent'] || 'Mozilla/5.0',
        'Accept': req.headers['accept'] || '*/*'
    };

    const protocol = options.protocol === 'https:' ? https : http;
    
    const proxyReq = protocol.request(options, (proxyRes) => {
        res.setHeader('Access-Control-Allow-Origin', '*');
        res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
        res.setHeader('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
        
        res.statusCode = proxyRes.statusCode;
        Object.keys(proxyRes.headers).forEach(key => {
            if (key.toLowerCase() !== 'access-control-allow-origin') {
                res.setHeader(key, proxyRes.headers[key]);
            }
        });

        proxyRes.pipe(res);
    });

    proxyReq.on('error', (err) => {
        console.error('Proxy error:', err.message);
        res.statusCode = 500;
        res.end('Proxy error: ' + err.message);
    });

    proxyReq.end();
}

const server = http.createServer((req, res) => {
    console.log(`${req.method} ${req.url}`);

    if (req.method === 'OPTIONS') {
        res.setHeader('Access-Control-Allow-Origin', '*');
        res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
        res.setHeader('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
        res.statusCode = 200;
        res.end();
        return;
    }

    const requestPath = req.url;
    let targetUrl = null;

    for (const [localPath, cdnBase] of Object.entries(CDN_MAPPINGS)) {
        if (requestPath.startsWith(localPath)) {
            const remainingPath = requestPath.substring(localPath.length);
            targetUrl = cdnBase + remainingPath;
            break;
        }
    }

    if (targetUrl) {
        forwardRequest(req, res, targetUrl);
    } else {
        res.statusCode = 404;
        res.end('CDN path not found');
    }
});

server.listen(PORT, () => {
    console.log(`CDN Proxy running on http://localhost:${PORT}`);
    console.log('Available: /wllama/ and /coi-serviceworker/');
});

module.exports = server; 

File: cdn_proxy.py
import http.server
import urllib.request
import urllib.parse

PORT = 8001

CDN_MAPPINGS = {
    '/wllama/': 'https://cdn.jsdelivr.net/npm/@wllama/wllama@2.3.2/esm/',
    '/wllama/single-thread/': 'https://cdn.jsdelivr.net/npm/@wllama/wllama@2.3.2/esm/single-thread/',
    '/wllama/multi-thread/': 'https://cdn.jsdelivr.net/npm/@wllama/wllama@2.3.2/esm/multi-thread/'
}

class CDNProxyHandler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        request_path = self.path
        target_url = None
        
        for local_path, cdn_base in CDN_MAPPINGS.items():
            if request_path.startswith(local_path):
                remaining_path = request_path[len(local_path):]
                target_url = cdn_base + remaining_path
                break
        
        if not target_url:
            self.send_error(404, "CDN path not found")
            return
        
        try:
            req = urllib.request.Request(target_url)
            req.add_header('User-Agent', 'Mozilla/5.0')
            
            with urllib.request.urlopen(req, timeout=30) as response:
                self.send_response(200)
                self.send_header('Access-Control-Allow-Origin', '*')
                self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
                self.send_header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept')
                self.send_header('Content-Type', response.headers.get('Content-Type', 'application/octet-stream'))
                self.end_headers()
                
                data = response.read()
                self.wfile.write(data)
                print(f"Proxied: {request_path} -> {target_url}")
                
        except Exception as e:
            print(f"Error: {e}")
            self.send_error(500, str(e))

def main():
    server_address = ('', PORT)
    httpd = http.server.HTTPServer(server_address, CDNProxyHandler)
    print(f"CDN Proxy running on http://localhost:{PORT}")
    print("Available: /wllama/ (wllama AI library endpoints)")
    httpd.serve_forever()

if __name__ == "__main__":
    main() 